# First glance at the dognition data 

**The main focus in this exercise is to explore any relationship between total tests completed and test itselt and membership type**

* Import dataset and take the first look at the data


* Look for a clue of what may affect total tests completed: 

 - start with the data of total tests completed itself: from which tests the users do not continue the test;
 
 - median of each test time: could it because users feel the tests are too easy (infered by a short test time) or the tests are too hard (infered by a long test time);
 
 - time difference between the first and last test (the idea was abandonded because the time diff totally depends the total tests);
 
 - After knowing the difference of membership type, we start to explore any pattern between membership type and total tests completed.


In [1]:
import pandas as pd

df_user = pd.read_excel('/Users/jingai/workspace/Capstone-project/Dognition-coursera/raw-data/dognition_user.xlsx')
df_dog= pd.read_excel('/Users/jingai/workspace/Capstone-project/Dognition-coursera/raw-data/dognition_dog.xlsx')

In [19]:
print(df_user.columns)
print(df_user.shape)
print(df_dog.columns)
print(df_dog.shape)

Index(['Created_at', 'Updated_at', 'User_ID', 'Dog_ID', 'Test_name',
       'Subcategory_name', 'Gender', 'Birthday', 'Breed', 'Breed_Type',
       'Breed_Group', 'Weight', 'Dog_Fixed', 'DNA_Tested', 'Dimension',
       'Sign_in_Count', 'Max_Dogs', 'Membership_ID', 'Subscribed', 'City',
       'State', 'Zip', 'Country', 'Exclude', 'Free_Start_User',
       'Last_Active_At', 'Membership_Type', 'Rating', 'Rank_by_UserID',
       'Rank_by_DogID'],
      dtype='object')
(177481, 30)
Index(['Dog ID', 'Total Tests Completed', 'Mean ITI (days)',
       'Mean ITI (minutes)', 'Median ITI (days)', 'Median ITI (minutes)',
       'Time diff between first and last game (days)',
       'Time diff between first and last game (minutes)', 'User ID', 'Gender',
       'Birthday', 'Breed', 'Breed_Type', 'Breed_Group', 'Weight', 'Dog_Fixed',
       'DNA_Tested', 'Dimension', 'Sign_in_Count', 'Max_Dogs', 'Membership_ID',
       'Subscribed', 'City', 'State', 'Zip', 'Country', 'Exclude',
       'Free_Start_U

## 1 Browsing a relation between total tests completed and tests time

In [ ]:
# for the whole population
df_dog[['Total Tests Completed', 'Median ITI (minutes)', 'Time diff between first and last game (days)']].describe().round(2)

In [ ]:
# for dogs who complete 20 tests
# filter out that Mean ITI and Time diff > 0.1 which does not make sense

filt_comp = (df_dog['Total Tests Completed'] >= 20) & (df_dog['Median ITI (minutes)'] > 0.1) & (df_dog['Time diff between first and last game (days)'] > 0.1)
dog_complete = df_dog.loc[filt_comp]
dog_complete[['Total Tests Completed', 'Median ITI (minutes)', 'Time diff between first and last game (days)']].describe().round(2)

In [ ]:
# take a look at the dogs who did not complete the first 20 tests.

filt_incomp = (df_dog['Total Tests Completed'] < 20)
dog_incomp = df_dog.loc[filt_incomp]
dog_incomp[['Total Tests Completed', 'Median ITI (minutes)', 'Time diff between first and last game (days)']].describe().round(2)

### 1.1 First findings and hypothesis

#### 1.1.1 Interesting findings

1. For all dogs (no data are excluded)

    - the total tests tompleted ranges between 1 and 45, with 7 as a median, indicating that majority of dogs have not complete the tests. <i><font color='blue'>Maybe a histgram would be helpful to check out when most of users stopped continuing the tests. </font></i>
    
    - Majority of dogs spent 5.61 min completing one test, as seen median('Median ITI(minutes)')
    
    - The median of Time diff between first and last game is 0.72 days. <i><font color='blue'>This data may not be useful as some dogs only complete a few games, which reasonably show a short time diff.</font></i>
    
    
2. For dogs who completed 20 tests (excluding 'Median ITI(minutes)' < 0.1 and 'Time diff' < 0.1)

    - Most dogs just completed 20 tests; <i><font color='blue'>Why did not they continue completing the games? Promotion? Free trial ends (A recent promotion provides 20 tests)?</font></i>
    
    - Most dogs in this group spent 6.35 min completing one test, which is not far from the results from the whole population;
    
    - The median of 'time diff' is 14.24 days.
    

3. For dogs who have not completed 20 tests (excluding 'Median ITI(minutes)' < 0.1 and 'Time diff' < 0.1)

    - Majority of dogs in this group only completed 4 games, as seen that both 25 and 50 quatile are 4;<i><font color='blue'>Why majority of dogs stopped at the fourth game? Free trial ends (provide the first 4 games)?</font></i>
    
    - Most dogs in this group spent 5.33 min completing one test, which is not far from the results from the whole population;
    
    - The median of 'time diff' is 0.03 days, indicating that most dogs/users might have given up after trying the first few games.
    

#### 1.1.2 Hypothesis and possible actions

The Hypothesis has been noted right after each interesting findings. 

In order to come up with recommedations on how to make more dogs/users complete tests, we need firstly figure out what could cause them stop completing tests, and which factors are linked with the number of total completed tests. The first glance shows a strong indication that most of users may stop the tests when free trials end. Thus, the next the step is to find out 

1. at which tests users do not continue the tests the most <i><font color='blue'>proper histogram</font></i> <i><font color='green'>(wait for finishing matlibplot tutorial video)</font></i>

2. take a look at 'Membership_Type' data and try to find a link with 'Total Tests Completed'



## 2 A first glance at membership type
    
**Membership type information**

1. ***Type 1** Membership: Dognition Assessment of initial 20 games*

2. ***Type 2** - Annual subscription: provide the Dognition Assessment plus 12 months of subscription service (where one new test and activity are “unlocked” per month, beginning of the month after purchase)*

3. ***Type 3** - Monthly subscription: provide the Dognition Assessment plus the subscription service billed monthly.*

4. ***Type 4** - Free subscriptions (the first 4 games for free) but needs to upgrade for the rest.*
    
5. ***Type 5** - Subscription: a recent test that offers the entire 20-game Dognition Assessment for free, but then allows user to upgrade to a monthly subscription.*

### 2.1 Any link between membership type and total completed tests?

In [8]:
df_dog['Membership_Type'].value_counts()

1     9129
4     3874
2     3735
3      659
0      502
5       46
CA      40
Name: Membership_Type, dtype: int64

The first glance show that majority of users assessed to the games by the Dognition Assessment of initial 20 games (membership **type 1**), following free subscriptions (**type 4**) which were offered to have access to the first games for free. The number of Annual subscriptions (**type 2**) are also noticeable. And only a small portion of uses chose Monthly subscriptions (**type 3**) and were offered the entire 20-game Dognition Assessment for free (**Type 5**). 

Type 0 and CA are not defined is the schema table, which are excluded for analysis.

Next, we are going to look at the total completed tests for each membership type group via `df.groupby()`.


In [15]:
# filter out the undefined membership type
filt_memb = ((df_dog['Membership_Type'] != 0) & (df_dog['Membership_Type'] != 'CA'))

#create a dataframegroup based on 'Membership_Type'
membership_grp = df_dog.loc[filt_memb].groupby(['Membership_Type'])

#Check whether we got the desired results
membership_grp['Membership_Type'].value_counts()

#Glance at the statistic description of 'Total Tests Completed' for each type of membership
membership_grp['Total Tests Completed'].apply(lambda x: x.describe().round(2))

Membership_Type       
1                count    9129.00
                 mean       10.82
                 std         6.86
                 min         1.00
                 25%         4.00
                 50%         9.00
                 75%        20.00
                 max        23.00
2                count    3735.00
                 mean       13.94
                 std         9.70
                 min         1.00
                 25%         4.00
                 50%        14.00
                 75%        20.00
                 max        45.00
3                count     659.00
                 mean       12.54
                 std         8.82
                 min         1.00
                 25%         4.00
                 50%        11.00
                 75%        20.00
                 max        45.00
4                count    3874.00
                 mean        3.93
                 std         2.08
                 min         1.00
                 25%     

### 2.2 First findings and preliminary conclusions

* Comparing median of total tests completed: **Type 2(14) > Type 3(11) = Type 5(11) > Type 1(9) > Type 4(4)** <font color="blue"><i>Histograms for each membership type on total tests completed would help.</i></font>

   
* The number of Annual subscribers are 3735 (which is a quite noticeable portion in total users), so the output can be reliable to some extent. The median of total tests completed is even higher than that of Type 5 (free for the initial 20 games). <font color="blue"><i>The preliminary thinking is that Annual subscribers tend to complete more tests probably because they care their dogs more.</i></font>


* Offering the first 20 tests free only show a slightly better effect on total tests completed compared to the membership who paid for the first 20 tests at once. <font color="blue"><i>As the counts of Type 5 is quite small (46), maybe the conclusion is not persuasive. More data needed?</i></font>


* Offering the first 4 tests for free trial showed the lowest median of total tests completed despite a large portion of users. <font color="blue"><i>It suggests that this promotion stragety may not help increasing the total tests completed.</i></font>

### 2.3 Any link between membership type and whether a user has a paid subscription or not and total tests completed?

In [60]:
#First glance at the whole population
paid_grp = df_dog.groupby(['Subscribed'])
paid_grp[['Total Tests Completed']].apply(lambda x: x.describe().round(2))

Total Tests Completed
Subscribed                             
0.0        count                8109.00
           mean                    7.30
           std                     6.45
           min                     1.00
           25%                     3.00
           50%                     4.00
           75%                     8.00
           max                    45.00
1.0        count                9876.00
           mean                   11.87
           std                     8.19
           min                     1.00
           25%                     4.00
           50%                    11.00
           75%                    20.00
           max                    45.00

#### 2.3.1 A rough conclusion on whole samples

* There are half of users who have paid subscriptions and half do not.

* In general, subscribed users completed more tests than non-subscribed users.

* As whether a user has a paid subscription highly depend on the type of members, we are gonna look at 'subscribed' cases in each membership type.

In [45]:
# how many users in each type of membership in total
membership_counts = df_dog.loc[filt_memb]['Membership_Type'].value_counts()
membership_counts

1    9129
4    3874
2    3735
3     659
5      46
Name: Membership_Type, dtype: int64

In [59]:
# how many users in each type of membership have a paid subscription
memb_paid_counts = membership_grp['Subscribed'].apply(lambda x: x.sum())
memb_paid_counts

Membership_Type
1    6111.0
2    3118.0
3     556.0
4      16.0
5      42.0
Name: Subscribed, dtype: float64

In [56]:
# combine above two tables into one table showing percentage
df_memb_paid_perc = pd.concat([membership_counts, memb_paid_counts],axis='columns', sort=False)

#Rename the column giving a reasonable name
df_memb_paid_perc.rename(columns={'Membership_Type': 'NumMemberType', 'Subscribed':'NumPaid'}, inplace=True)

#Add a new column showing percentage
df_memb_paid_perc['PercMembPaid'] = (df_memb_paid_perc['NumPaid']/df_memb_paid_perc['NumMemberType']*100).round(2)

#Add a new column showing the median of total tests completed
median_tot_tests = membership_grp['Total Tests Completed'].apply(lambda x: x.median())
df_memb_paid_perc['MedianTotTests'] = median_tot_tests

#Sort by median of total tests completed
df_memb_paid_perc = df_memb_paid_perc.sort_values(by='MedianTotTests', ascending=False)
df_memb_paid_perc

NumMemberType  NumPaid  PercMembPaid  MedianTotTests
2           3735   3118.0         83.48            14.0
3            659    556.0         84.37            11.0
5             46     42.0         91.30            11.0
1           9129   6111.0         66.94             9.0
4           3874     16.0          0.41             4.0

#### 2.3.2 Rough conclusions on each type of membership

* A quick look does not show any patten between paid subscription and total tests completed in each group. But in general, more tests completed are found in the group with a higher percentage of paid subscription. <font color='blue'><i>Maybe try to encourage users to become a paid subscription would help to compete more tests since people are more likely to enjoy the service if they are paying. </i></font>


* It is interesting to see that only very few users have a paid subscription if they access to the Dognition tests through a free trial for the first 4 tests. This again indicates that **offering a 4-tests free trial may not be a good idea to encourage users to complete the tests**.


## 3 What the common feature of users who completed tests

### 3.1 Common features of dog owners in each membership type

As users who are annual and monthly members tend to complete more games, which we assume that it is because those owners care about their dogs more, we are gonna find proof for it. In other words, we are trying to find which dog owners are more likely to go annual subscription.

#### 3.1.1 Start with whether dogs are fixed, have DNA tested and how many dogs the owners have.

In [74]:
membgrp_dog=membership_grp[['Dog_Fixed','DNA_Tested']].apply(lambda x: x.sum())
membgrp_dog

Dog_Fixed  DNA_Tested
Membership_Type                       
1                   8032.0       218.0
2                   3225.0       258.0
3                    541.0         6.0
4                   3033.0       175.0
5                     38.0         0.0

In [76]:
perc_dnatested

1    2.387994
2    6.907631
3    0.910470
4    4.517295
5    0.000000
dtype: float64

In [78]:
perc_fixed = (membgrp_dog['Dog_Fixed']/membership_counts*100).round(2)
perc_dnatested = (membgrp_dog['DNA_Tested']/membership_counts*100).round(2)

df_memb_dog=pd.concat([perc_fixed, perc_dnatested, median_tot_tests], axis=1, sort=False)

df_memb_dog.rename(columns={0: 'PercFix', 1: 'Percdnatested', 'Total Tests Completed': 'MedianTotTests'}, inplace=True)
df_memb_dog


PercFix  Percdnatested  MedianTotTests
1    87.98           2.39             9.0
2    86.35           6.91            14.0
3    82.09           0.91            11.0
4    78.29           4.52             4.0
5    82.61           0.00            11.0

#### 3.1.2 First findings 

* No clearn pattern can be found between percentage of dogs fixed and membership type. So **whether the dog is fixed or not might not matter**.

* It seems that the users with Annual subscriptions showed the highest percentage of dogs got DNA tested. <font color='blue'><i>Could it because these users own a pure breed dog? I.e., that's why they care their dogs more and want to do assessment on their dogs?</i></font>

To affirm above hypothesis, we are gonna take a look at the breed type of the dogs in each membership type.

#### 3.1.3 DNA tests and breed type

In [79]:
df_dog['Breed_Type'].value_counts()

Pure Breed                          9346
Mixed Breed/ Other/ I Don't Know    4999
Cross Breed                         2976
Popular Hybrid                       664
Name: Breed_Type, dtype: int64

In [94]:
# count how many pure breed dogs in each membership type
pure_breed_counts = membership_grp['Breed_Type'].apply(lambda x: x.str.contains('Pure Breed').sum())
hybrid_counts = membership_grp['Breed_Type'].apply(lambda x: x.str.contains('Popular Hybrid').sum())
cross_counts = membership_grp['Breed_Type'].apply(lambda x: x.str.contains('Cross Breed').sum())
mixed_counts = membership_grp['Breed_Type'].apply(lambda x: x.str.contains('Mixed Breed').sum())

perc_pure_breed = (pure_breed_counts/membership_counts*100).round(2)
perc_hybrid = (hybrid_counts/membership_counts*100).round(2)
perc_cross = (cross_counts/membership_counts*100).round(2)
perc_mixed = (mixed_counts/membership_counts*100).round(2)

df_memb_dog['PercPure'], df_memb_dog['PercHybrid'], df_memb_dog['PercCross'],df_memb_dog['PercMix'] = [perc_pure_breed, perc_hybrid, perc_cross, perc_mixed]

df_memb_dog


PercFix  Percdnatested  MedianTotTests  PercPure  PercHybrid  PercCross  \
1    87.98           2.39             9.0     49.57        4.25      18.03   
2    86.35           6.91            14.0     57.43        3.83      14.24   
3    82.09           0.91            11.0     52.50        3.49      16.39   
4    78.29           4.52             4.0     51.70        2.35      15.38   
5    82.61           0.00            11.0     65.22        2.17      13.04   

   PercMix  
1    28.15  
2    24.50  
3    27.62  
4    30.56  
5    19.57

#### 3.1.4 Rough observations on DNA tests and breed type of dogs

If we exclude the membership type 5 (as the number of samples is too small), users who have annual subscriptions do show the highest percentage of pure breed dogs. It sort of corroborate our hypothesis: 

<font color='red'><b>Users who own pure breed dogs tend to know more about their dogs' personality and dimensions</b></font>